Tasks/Common/ios-signing-common/ios-signing-common.ts

```diff
@@ -602,8 +598,83 @@ export async function getP12CommonName(p12Path: string, p12Pwd: string) {
     return commonName;
 }
 
-export async function deleteCert(keychainPath: string, certSha1Hash: string) {
+export async function deleteCert(keychainPath: string, certSha1Hash: string): Promise<void> {
     let deleteCert: ToolRunner = tl.tool(tl.which('security', true));
     deleteCert.arg(['delete-certificate', '-Z', certSha1Hash, keychainPath]);
     await deleteCert.exec();
 }
+
+export async function getP12PrivateKeyName(p12Path: string, p12Pwd: string): Promise<string> {
+    //openssl pkcs12 -in <p12Path> -nocerts -passin pass:"<p12Pwd>" -passout pass:"<p12Pwd>" | grep 'friendlyName'
+    //We pipe through grep so we we don't log the private key, even encrypted.
+    tl.debug('getting the P12 private key name');
+    let opensslPath: string = tl.which('openssl', true);
+    let openssl: ToolRunner = tl.tool(opensslPath);
+    if (!p12Pwd) {
+        // if password is null or not defined, set it to empty
+        p12Pwd = '';
+    }
+    openssl.arg(['pkcs12', '-in', p12Path, '-nocerts', '-passin', 'pass:' + p12Pwd, '-passout', 'pass:' + p12Pwd]);
+
+    let grepPath: string = tl.which('grep', true);
+    let grep: ToolRunner = tl.tool(grepPath);
+    grep.arg(['friendlyName']);
+    openssl.pipeExecOutputToTool(grep);
+
+    let privateKeyName: string;
+    openssl.on('stdline', function (line: string) {
+        if (line) {
+            const match = line.match(/friendlyName: (.*)/);
+            if (match && match[1]) {
+                privateKeyName = match[1].trim();
+            }
+        }
+    });
+
+    await openssl.exec();
+    tl.debug('P12 private key name = ' + privateKeyName);
+    if (!privateKeyName) {
+        throw tl.loc('P12PrivateKeyNameNotFound', p12Path);
+    }
+
+    return privateKeyName;
+}
+
+/**
+ * Set the partition_id ACL so codesign has permission to use the signing key.
+ */
+async function setKeyPartitionList(keychainPath: string, keychainPwd: string, privateKeyName: string) {
+    // security set-key-partition-list -S apple: -s -l <privateKeyName> -k <keychainPwd> <keychainPath>
+    // n.b. This command could update multiple keys (e.g. an expired signing key and a newer signing key.)
+
+    if (privateKeyName) {
+        tl.debug(`Setting the partition_id ACL for ${privateKeyName}`);
+
+        // "If you'd like to run /usr/bin/codesign with the key, "apple:" must be an element of the partition list." - security(1) man page.
+        // When you sign into your developer account in Xcode on a new machine, you get a private key with partition list "apple:". However
+        // "security import a.p12 -k login.keychain" results in the private key with partition list "apple-tool:".
+        const partitionList = 'apple:';
```

Comment:
Most people (see stackoverflow, fastlane) retain the `apple-tool:` that you get when using `security import`. I'm inclined to leave it off to mimic Xcode's behavior when adding a private key for the developer.

BTW, before I tested Xcode's behavior, I was originally going to run `security dump-keychain -a` to get the existing ACLs so I could add `apple:` to them. However `dump-keychain -a login.keychain` takes 3 minutes on the Mac mini and 13 minutes on my MacBook Pro!

Reply:
Done.